### PR TITLE
move linkerd-web as a add-on

### PR DESCRIPTION
**Pull request by Pothulapati**
This PR moves linkerd-web into a separate chart called dashboard and
thus making it optional. This follows the same approach as that of
grafana and prometheus.

Signed-off-by: Tarun Pothulapati <tarunpothulapati@outlook.com>

### CODE REVIEW

File: cli/cmd/install.go [hunk 1]
```diff
@@ -791,7 +789,6 @@ func (options *installOptions) buildValuesWithoutIdentity(configs *pb.All) (*l5d
 	installValues.HeartbeatSchedule = options.heartbeatSchedule()
 	installValues.RestrictDashboardPrivileges = options.restrictDashboardPrivileges
 	installValues.DisableHeartBeat = options.disableHeartbeat
-	installValues.WebImage = fmt.Sprintf("%s/web", options.dockerRegistry)
```
Comment (Pothulapati) [inline cli/cmd/install.go:794]: This image name can be overriden by using `--addon-config` and through `values.yaml` with Helm. but when a user applies the `--registry` flag, Is the expectation that the image registry value is overriden even for add-ons? We can make sure its propagated to add-ons if thats the case. @alpeb

File: charts/add-ons/dashboard/templates/web.yaml [hunk 2]
```diff
@@ -68,12 +68,10 @@ spec:
...
         - -grafana-addr={{.Values.global.grafanaUrl}}
-        {{- else if .Values.grafana.enabled }}
-        - -grafana-addr={{.Values.grafana.name}}.{{.Values.global.namespace}}.svc.{{.Values.global.clusterDomain}}:3000
-        {{- end}}
-        {{- if .Values.tracing.enabled }}
-        - -jaeger-addr={{.Values.tracing.jaeger.name}}.{{.Values.global.namespace}}.svc.{{.Values.global.clusterDomain}}:16686
+        {{- else}}
+        - -grafana-addr=linkerd-grafana.{{.Values.global.namespace}}.svc.{{.Values.global.clusterDomain}}:3000
         {{- end}}
+        - -jaeger-addr=linkerd-jaeger.{{.Values.global.namespace}}.svc.{{.Values.global.clusterDomain}}:16686
```
Comment (Pothulapati) [inline charts/add-ons/dashboard/templates/web.yaml:74]: but as we are removing the condition and passing that variable all the time, This means that users will see the `jaeger` icon irrespective of the add-on being installed which is not great. wondering if there's a better way to do this. :thinking: 

@alpeb

File: cli/cmd/install.go (hunk 1 above)
Comment (alpeb) [reply cli/cmd/install.go:794]: Yeah I guess that is the expectation, for `--registry` to apply to all the images, including add-ons.

File: charts/add-ons/dashboard/templates/web.yaml (hunk 2 above)
Comment (alpeb) [reply charts/add-ons/dashboard/templates/web.yaml:74]: This stems from the inability of add-ons to access each other's Helm configs, right?
The web component currently mounts `linkerd-config`; we could have it mount `linkerd-config-addons` as well, and override the grafana/jaeger flags based off of that.

Comment (Pothulapati) [reply charts/add-ons/dashboard/templates/web.yaml]: @alpeb Great Suggestion!
Updated the PR regarding the same!

File: web/main.go [hunk 3]
```diff
@@ -57,6 +59,33 @@ func main() {
...
 
+	values, err := config.AddOns(pkgK8s.MountPathAddOnsConfig)
+	if err != nil {
+		log.Warnf("failed to load values config: [%s]", err)
+	}
+
+	addOns, err := l5dcharts.ParseAddOnValues(values)
+	if err != nil {
+		log.Warnf("failed to parse add-on values: [%s]", err)
+	}
```
Comment (alpeb) [inline web/main.go:70]: How do you feel about moving this logic into `config.Addons()`?

File: web/main.go [hunk 4]
```diff
@@ -57,6 +59,33 @@ func main() {
...
+
+	for _, addOn := range addOns {
+		if addOn.Name() == l5dcharts.GrafanaAddOn && *grafanaAddr == "" {
+			*grafanaAddr = fmt.Sprintf("linkerd-grafana.%s.svc.%s:3000", *controllerNamespace, clusterDomain)
+		}
+
+		if addOn.Name() == l5dcharts.TracingAddOn && *jaegerAddr == "" {
+			*jaegerAddr = fmt.Sprintf("linkerd-jaeger.%s.svc.%s:16686", *controllerNamespace, clusterDomain)
```
[the remark this answers is not in the document]
Comment (Pothulapati) [reply web/main.go:78]: `l5dcharts.ParseAddOnValues()` actually gives back a list of **enabled** add-ons. So, No need to check if they are enabled again.

[the remark this answers is not in the document]
Comment (alpeb) [reply web/main.go:78]: :+1:

**Comment by Pothulapati** [comment]
@alpeb Yep, That was my plan!

**Comment by Pothulapati** [comment]
Minimal Control Plane effort is being worked as a separate effort (with an RFC) along with a slightly different approach. So, Closing this! See #4612

**Pull request closed without merging**